```{r}
library(tidyverse)
library(tidymodels)
library(here)
library(reticulate)
library(vroom)
library(DarkKinaseTools)
library(fastshap)
library(conflicted)
conflict_prefer("rename", "dplyr")
conflict_prefer("filter", "dplyr")
conflict_prefer("explain", "fastshap")
#py_install("shap")
#py_install("matplotlib")
# matplotlib <- import("matplotlib")
# matplotlib$use("Agg", force = TRUE)
```

```{r}
#read in data 
data = read_rds(here('results/PRISM_LINCS_klaeger_models_auc/PRISM_LINCS_klaeger_data_for_ml_5000feat_auc.rds.gz'))

cors =  vroom(here('results/PRISM_LINCS_klaeger_models_auc/PRISM_LINCS_klaeger_data_feature_correlations_auc.csv'))

final_model = read_rds(here('results/PRISM_LINCS_klaeger_models_auc/activation_expression/regression/xgboost/final_xgboost_model.rds.gz'))

PDAC_dataset_for_model = read_rds(here('results/validation_results/PDAC_validation/PDAC_data_for_model.rds'))

gene_classifications = read_csv(here('data/KlaegerScience2017/manual_gene_labels.csv')) %>%
	bind_rows(all_kinases %>% select(symbol,class) %>% rename(gene_name = symbol)) %>% 
	#adding extra label for abbreviated activation name
	bind_rows(data.frame(gene_name = "CSNK2A(1/3)", class = "Light")) %>%
	distinct() %>% 
	mutate(gene_name = str_replace(gene_name, "[-;]", "_")) %>% 
	mutate(gene_name = if_else(
		gene_name == 'HIST2H2BE_HIST1H2BB;HIST1H2BO;HIST1H2BJ;HIST3H2BB;HIST1H2BA',
		'HIST2H2BE_HIST1H2BB_HIST1H2BO_HIST1H2BJ_HIST3H2BB_HIST1H2BA',
		gene_name
	))

exp_interactor_type = read_rds(here('results/PRISM_LINCS_klaeger_models_auc/expression_feature_interactors.rds')) %>%
	mutate(Variable = paste0(type,"_",hgnc)) %>%
	mutate(interaction = case_when(
		direct_count >= 1 ~ "Inhibition\nInteractor",
		T ~ "Noninteractor"
	)) 
```

```{r}
#define recipe
build_all_data_regression_viability_set = function(num_features, all_data, feature_correlations) {
	this_data_filtered = all_data %>%
		select(any_of(feature_correlations$feature[1:num_features]),
					 depmap_id,
					 ccle_name,
					 broad_id,
					 auc)
}
this_dataset = build_all_data_regression_viability_set(feature_correlations =  cors,
																													 num_features = 5000,
																													 all_data = data)

this_recipe = recipe(auc ~ ., this_dataset) %>%
	update_role(-starts_with("act_"),
							-starts_with("exp_"),
							-starts_with("auc"),
							new_role = "id variable") 
```

```{r}
#preprocess PDAC data
PDAC_X = prep(this_recipe, 
							PDAC_dataset_for_model %>%
								mutate(broad_id = NA,
											 depmap_id = NA,
											 ccle_name = NA,
											 auc = NA,
											 exp_AL162231.1 = 0,
											 exp_AC093512.2 = 0,
											 exp_AC008397.1 = 0,
											 exp_AC007906.2 = 0,
											 exp_AL445238.1 = 0,
											 exp_AL445989.1 = 0)
							)%>% 
	juice() %>% 
	#write into csv for python
	write_csv(here('results/PRISM_LINCS_klaeger_models_auc/PDAC_data_for_shap.csv'))
```

```{r}
#shap values from python
shap_values = read_csv(here('results/PRISM_LINCS_klaeger_models_auc/PRISM_LINCS_klaeger_auc_xgboost_shap_values_python_PDAC.csv'))
shap_samples = read_csv(here('results/PRISM_LINCS_klaeger_models_auc/PRISM_LINCS_klaeger_samples_for_shap_values_python_PDAC.csv'))
```

```{r}
shap_values_individual = shap_values %>% 
	rename(row_number = `...1`) %>%
	mutate(row_number = row_number()) %>% 
	pivot_longer(-row_number, names_to = "feature", values_to = "feature_importance_value")

shap_samples_individual = shap_samples %>% 
	rename(row_number = `...1`) %>%
	mutate(row_number = row_number()) %>% 
	pivot_longer(-row_number, names_to = "feature", values_to = "feature_value")

shap_values_processed_individual = shap_values_individual %>%
	arrange((feature_importance_value)) %>%
	left_join(shap_samples_individual) %>% 
	mutate(feature_type = case_when(
			str_detect(feature, "^act_") ~ "Inhibition",
			str_detect(feature, "^exp_") ~ "Expression",
			str_detect(feature, "^dep_") ~ "Depmap",
			str_detect(feature, "^cnv_") ~ "CNV",
			str_detect(feature, "^prot_") ~ "Proteomics",
			T ~ feature
		))
	
shap_values_processed = shap_values_individual %>% 
	group_by(feature) %>% 
	summarise(mean_importance_value = mean(abs(feature_importance_value))) %>% 
	arrange((mean_importance_value)) %>%
	mutate(feature_type = case_when(
			str_detect(feature, "^act_") ~ "Inhibition",
			str_detect(feature, "^exp_") ~ "Expression",
			str_detect(feature, "^dep_") ~ "Depmap",
			str_detect(feature, "^cnv_") ~ "CNV",
			str_detect(feature, "^prot_") ~ "Proteomics",
			T ~ feature
		))

write_csv(shap_values_processed_individual, here('results/PRISM_LINCS_klaeger_models_auc/all_feature_shap_values_xgboost_PDAC.csv'))
```

```{r}
#separate by feature type and plot 
#shap_values_processed = read_csv(here('results/PRISM_LINCS_klaeger_models_auc/all_feature_shap_values_xgboost.csv'))

shap_inhibition = shap_values_processed %>% 
	filter(feature_type == "Inhibition") %>% 
	mutate(feature = ifelse(feature == "act_CSNK2A1_CSNK2A3","act_CSNK2A(1/3)",feature)) %>%
	extract(feature, c(NA,"gene"),"(.*)_(.*)",remove=F) %>%
	left_join(gene_classifications, by=c('gene'='gene_name')) %>%
	mutate(class = ifelse(is.na(class), "Non-kinase",class)) %>% 
	arrange(desc(mean_importance_value)) 

shap_expression = shap_values_processed %>% 
	filter(feature_type == "Expression") %>% 
	extract(feature, c(NA,"gene"),"(.*)_(.*)",remove=F) %>%
	left_join(exp_interactor_type %>% select(Variable,interaction) %>% rename(feature = Variable)) %>% 
	arrange(desc(mean_importance_value))
	
shap_inhibition %>%
  slice(1:25) %>%
	arrange(mean_importance_value) %>% 
	mutate(gene = fct_inorder(gene)) %>% 
	ggplot(aes(x = mean_importance_value, y = gene, fill = class)) +
	geom_col() +
	labs(y='',fill='',x='Inhibition State Importance') +
	theme(
		legend.title = element_blank(),
		legend.position = "bottom",
		legend.direction = "horizontal",
		legend.margin = margin(t = 0, b = 0, l = 0, r = 0, unit = "cm"),
		panel.background = element_rect(fill = "transparent",colour = NA),
    panel.grid.minor = element_blank(), 
    panel.grid.major = element_blank(),
    plot.background = element_rect(fill = "transparent",colour = NA)
      )
ggsave(here('figures/PRISM_LINCS_klaeger/inhibition_shap_figure_PDAC.png'), height = 10.5, width = 8.5, units = "cm")

shap_expression %>%
	mutate(interaction = if_else(gene == "KCTD14" | gene == "CARMIL2" | gene == "ASAP2",
															 "Noninteractor",
															 interaction)) %>% 
  slice(1:25) %>%
	arrange(mean_importance_value) %>% 
	unique() %>% 
	mutate(gene = fct_inorder(gene)) %>% 
	ggplot(aes(x = mean_importance_value, y = gene, fill = interaction)) +
	geom_col() +
	labs(y='',fill='',x='Gene Expression Importance') +
	#scale_x_continuous(breaks = c(0, 5e-4,1e-3)) +
	theme(
		legend.title = element_blank(),
		legend.position = "bottom",
		legend.direction = "horizontal",
		legend.margin = margin(t = 0, b = 0, l = 0, r = 0, unit = "cm"),
		panel.background = element_rect(fill = "transparent",colour = NA),
    panel.grid.minor = element_blank(), 
    panel.grid.major = element_blank(),
    plot.background = element_rect(fill = "transparent",colour = NA)
      )
ggsave(here('figures/PRISM_LINCS_klaeger/expression_shap_figure_PDAC.png'), height = 10.5, width = 8.5, units = "cm")
```

```{r}
#force summary plots
#separate by feature type and plot 
shap_inhibition_individual = shap_values_processed_individual %>% 
	filter(feature_type == "Inhibition") %>% 
	mutate(feature = ifelse(feature == "act_CSNK2A1_CSNK2A3","act_CSNK2A(1/3)",feature)) %>%
	extract(feature, c(NA,"gene"),"(.*)_(.*)",remove=F) %>%
	left_join(gene_classifications, by=c('gene'='gene_name')) %>%
	mutate(class = ifelse(is.na(class), "Non-kinase",class)) %>% 
	arrange(desc(feature_importance_value))

shap_inhibition_order = shap_inhibition_individual %>% 
	group_by(feature, feature_type, gene, class) %>% 
	summarise(feature_importance_value = mean(abs(feature_importance_value))) %>% 
	ungroup() %>% 
	arrange(feature_importance_value) %>% 
	slice_tail(n = 25)
	
shap_inhibition_individual %>%
  filter(feature %in% shap_inhibition_order$feature) %>%
	mutate(feature_value = 
				 	if_else(
				 		feature_value > 1,
				 		1,
				 		feature_value
				 	)) %>% 
	mutate(feature_value = 100*(1 - feature_value)) %>% 
	mutate(gene = as.factor(gene)) %>% 
	mutate(gene = fct_relevel(gene, shap_inhibition_order$gene)) %>% 
	ggplot(aes(y = gene, x = feature_importance_value, colour = feature_value)) +
	geom_jitter(position = "dodge") +
	scale_colour_viridis_c() +
	#coord_flip() +
	labs(y='',fill='',x='Kinase Inhibition Impact on Cell Line Response', colour = "Kinase Inhibition Level (%)") +
	theme(
		#legend.title = element_blank(),
		legend.position = "bottom",
		legend.direction = "horizontal",
		legend.margin = margin(t = 0, b = 0, l = 0, r = 0, unit = "cm"),
		panel.background = element_rect(fill = "transparent",colour = NA),
    panel.grid.minor = element_blank(), 
    panel.grid.major = element_blank(),
    plot.background = element_rect(fill = "transparent",colour = NA)
      )
ggsave(here('figures/PRISM_LINCS_klaeger/inhibition_shap_summary_figure_PDAC.png'), height = 10.5, width = 8.5, units = "cm")

shap_inhibition_order %>% 
	mutate(gene = as.factor(gene)) %>% 
	mutate(gene = fct_relevel(gene, shap_inhibition_order$gene)) %>%
	ggplot(aes(x = feature_type, y = gene, fill = class)) +
	geom_tile() +
	#scale_fill_viridis_d() +
  theme(
  	legend.position = "left",
  	legend.direction = "vertical",
  	#axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
  	axis.title.x = element_blank(),
  	axis.ticks.x = element_blank(),
  	#axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
  	legend.margin = margin(t = 0, b = 0, l = 0, r = 0, unit = "cm"),
		panel.background = element_rect(fill = "transparent",colour = NA),
    panel.grid.minor = element_blank(), 
    panel.grid.major = element_blank(),
    plot.background = element_rect(fill = "transparent",colour = NA))
	
ggsave(here('figures/PRISM_LINCS_klaeger/inhibition_shap_summary_figure_annotations_PDAC.png'), height = 10.5, width = 8.5, units = "cm")
```

